### DISCUSSION TITLE
如何传递自定义工具链设置

### DISCUSSION THREAD

**Discussion by eddylin83** [xmake-io/xmake#2789] [Q&A]
### 你在什么场景下需要该功能？

在本机+交叉编译复合项目中，包含了一个子工程。如果在不修改子工程的情况下，设置其交叉编译工具链？
```
includes("xlib") -- 包含一个子工程代码

target("native")
    set_kind("binary")
    set_plat(os.host())
    set_arch(os.arch())
    add_files("utils/src/*.c")

target("cross")
    set_kind("binary")
    set_toolchains("cross-tools")
    add_deps("xlib")
    add_files("board/src/*.c")
```
如果在全局进行set_toolchains可以传递进去，但是又会影响native这个目标。

### 描述可能的解决方案

用rule或者option能否实现类似功能，或者本来就有简单的实现方式？

**Comment by waruqi** [comment]
> 如果在全局进行set_toolchains可以传递进去，但是又会影响native这个目标。

重新组织下 includes 结构，native 挪到顶层，cross / xlibs 放到子 xmake.lua ，然后子 xmake.lua 全局配置 set_toolchains

**Comment by waruqi** [comment]
rules 可以，全局add_rules，然后 on_load 里面自己判断要对哪些 target 设置 toolchains
